text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+addressed.)
+
+Although I am hesitant to make a firm assessment without doing some more
+real-world measurement and benchmarking, my hunch is that we should be able to
+create a memory-safe API and user-friendly API with roughly equivalent
+performance to existing C++ implementations.
+
+### zerocopy mutability
+
+Adding mutability to this design poses challenges. Fundamentally, the problem
+comes down to Rust's memory safety guarantees. Specifically, Rust forbids
+aliasing mutable references. In the zerocopy case, we would want to be able to
+have mutable references to multiple tables at the same time. These tables share
+the same underlying storage, so we cannot just pass out multiple mutable
+references to that storage to our multiple tables.
+
+The zerocopy crate *is* designed with this in mind: this is the reason for the
+API of `LayoutVerified`, which can read off the front of a slice: under the
+covers, it is splitting the pointer, ensuring that it has exclusive access. This
+pattern may cover *some* of our needs, but not all of them, and even where it
+was feasible it would add significant complexity. With the top-level
+tables, for instance, we could verify when loading the font that the reported
+offsets for the tables were all disjoint, and then we could split the underlying
+data into chunks, each of which would be independently mutable; we could either
+do this explicitly (allocating an array of disjoint slices) or via some other
+mechanism, similar to a `RefCell`, where we would keep track at runtime of which regions
+were currently borrowed.
+
+This would work (although it would add significant complexity) in cases where we
+had foreknowledge of the position and size of an object, but I do not believe
+this could work when dealing with tables that host offsets, or at least not in
+the general case. The problem with offsets is that they are not verified ahead
+of time, and they often point to objects that do not have a known size. This
+means there is really no way to know exactly how much memory a table at a given
+offset is going to access. While it is possible to imagine complicated runtime
+schemes for tracking this, it would definitely require allocating additional
+data structures, and would significantly degrade the ergonomics of the API, and
+significantly increase the learning curve.
+
+I've been thinking about this problem quite a bit, and my opinion it is evidence
+of a fundamental tension in the design goals of this project.
+
+#### deciding on goals
+
+I believe that we currently have two conflicting goals. One of the rationale for
+this project is concern about the difficulty of getting developers up to speed
+on HarfBuzz, and how hard it is to add features there without accidentally
+introducing bugs. There is a hope that Rust's safety guarantees can give us a
+code base that is easier to modify and hack on; however there is *also* a desire
+to end up with code that is more-or-less equivalent to the existing C++.
```

Comment:
I think "code more or less equivalent" isn't quite right, further details in comment below.